### PR TITLE
Update to fsnotify `0.4+`

### PR DESCRIPTION
**Pull request by lua-vr**
This fixes some issues for me related to saving ~large files (like 100kb) with Emacs, like Vim it seems to have a weird way of writing (first it writes an empty file then fill it up, apparently). The previous hfsnotify's debouncer would only report the first write and the file would appear empty sometimes.

### CODE REVIEW

**Comment by lua-vr** [comment]
Just a note (about intentation diffs like last time): you can hide whitespace changes in diffs with `?w=1` at the end of the github URL, like so: https://github.com/srid/unionmount/pull/3/commits/2ba9b55a8b72a39bd6f99403f3c9500489f07f42?diff=split&w=1 (or `git diff -w` locally)

**Comment by srid** [comment]
Could you rebase from new `master` (so CI will run)? Version should be `0.2.1` I think. I'll test this with Emanote latter.

**Comment by lua-vr** [comment]
@srid I noticed the debounce I wrote was relying on the fact that the TBQueue had size 1, because to compare the last item with the new item it read from the queue. If the queue had size two, FIFO means the read item would not be the last anymore!

So I think the code should either make it clear the queue has size 1 by using a TMVar or keep using TBQueues but with larger size, as in this branch:
https://github.com/srid/unionmount/compare/master...lucasvreis:unionmount:tbqueue-proper?w=1

In the end I think both are more or less equivalent because `watchTreeM` (from fsnotify) keeps it's own internal queue of events anyway, so there should be no risk of missing actions due to the size of unionmount queue.

**Comment by srid** [comment]
CI is only failing because we probably need Nix overrides (I can do that, no problem):

image

**Comment by srid** [comment]
It'll be a while before I can review the diff. But I'm thinking that it would be good have tests (separate PR merged) so that PRs like this can be confidently reviewed and merged. Right now, I have to manually test this on emanote, and try to understand the code thoroughly (which should always be done anyway; but having automated tests increases the confidence).

EDIT: Perhaps I should look into nixos VM tests ...

File: src/System/UnionMount.hs [hunk 1]
```diff
@@ -180,33 +177,41 @@ unionMount' sources pats ignore = do
...
-                              changes <- fmap snd . flip runStateT Map.empty $ do
-                                put =<< lift . changeInsert src tag fp act =<< get
-                              lift $ reportChange changes
-                              loop
-                loop
+              let readDebounced = do
+                    -- Wait for some initial action in the queue.
+                    _ <- atomically $ readTMVar q
+                    -- 100ms is a reasonable wait period to gather (possibly related) events.
+                    liftIO $ threadDelay 100000
```
Comment (lua-vr) [inline src/System/UnionMount.hs:187]: This debounce interval is the same as before, but perhaps it should be made configurable

**Comment by maralorn** [comment]
I encountered this issue while we are trying to update nixpkgs to LTS 21.
If we can do a quick release of this I we can proceed without pinning fsnotify to an older version.

**Review by srid — APPROVED** [review]
Haven't taken a *detailed* look at this PR, but it works well in Emanote (confirmed in https://github.com/srid/emanote/pull/454). I trust your changes @lucasvreis

**Review by srid — APPROVED** [review]

**Comment by lua-vr** [comment]
FWIW, I've been running my ema app using this branch since May with no issues.

**Pull request merged**